File: apps/api/src/routes/relays.ts
import { Router, type Request, type Response } from "express";
import {
  insertCommand,
  getDeviceActuators,
  getDevice,
  updateActuatorState,
  updateActuatorName,
  removeActuatorName,
} from "../lib/sqlite.js";
import { publishCommand } from "../services/mqttTelemetry.js";
import { broadcastCommand } from "../services/websocket.js";

export function createRelaysRouter(): Router {
  const router = Router({mergeParams: true});

  // GET /api/devices/:deviceId/relays - List all relays from device actuators
  router.get("/", (req: Request, res: Response) => {
    try {
      const actuators = getDeviceActuators(String(req.params.deviceId));

      const relays = actuators.map((actuator) => ({
        id: actuator.id,
        name: actuator.customName ?? actuator.name ?? actuator.id,
        type: actuator.type ?? "switch",
        state: actuator.state ?? false,
        updatedAt: Date.now(),
        deviceId: actuator.deviceId,
        location: actuator.location,
        deviceOnline: actuator.deviceOnline,
      }));

      res.json({ ok: true, relays });
    } catch (err) {
      console.error("Error fetching relays", err);
      res.status(500).json({ ok: false, error: "Failed to fetch relays" });
    }
  });

  // GET /api/devices/:deviceId/relays/:id - Get a single relay by ID
  router.get("/:id", (req: Request, res: Response) => {
    try {
      const relayId = req.params.id as string;
      const actuators = getDeviceActuators(String(req.params.deviceId));
      const actuator = actuators.find((a) => a.id === relayId);

      if (!actuator) {
        res.status(404).json({ ok: false, error: "Relay not found" });
        return;
      }

      const relay = {
        id: actuator.id,
        name: actuator.customName ?? actuator.name ?? actuator.id,
        type: actuator.type ?? "switch",
        state: actuator.state ?? false,
        deviceId: actuator.deviceId,
        location: actuator.location,
        deviceOnline: actuator.deviceOnline,
      };

      res.json({ ok: true, relay });
    } catch (err) {
      console.error("Error fetching relay", err);
      res.status(500).json({ ok: false, error: "Failed to fetch relay" });
    }
  });

  // POST /api/devices/:deviceId/relays/:id - Control relay state (toggle on/off or pulse)
  router.post("/:id", (req: Request, res: Response) => {
    try {
      const relayId = req.params.id as string;
      const targetDeviceId = String(req.params.deviceId);

      // Find the actuator from device capabilities
      const actuators = getDeviceActuators(targetDeviceId);
      const actuator = actuators.find((a) => a.id === relayId);

      if (!actuator) {
        res.status(404).json({ ok: false, error: "Relay not found" });
        return;
      }

      const relayName = actuator.customName ?? actuator.name ?? relayId;
      const targetLocation = actuator.location;

      if (actuator.type === "momentary") {
        // Momentary: send pulse action, no optimistic state update
        const correlationId = publishCommand({
          deviceId: targetDeviceId,
          location: targetLocation,
          target: relayId,
          action: "pulse",
          value: true,
          source: "dashboard",
          reason: `Relay ${relayName} pulsed via dashboard`,
        });

        if (!correlationId) {
          res.status(503).json({ ok: false, error: "MQTT client not connected" });
          return;
        }

        const command = insertCommand({
          id: correlationId,
          ts: Date.now(),
          deviceId: targetDeviceId,
          target: relayId,
          action: "pulse",
          value: true,
          source: "dashboard",
          reason: `Relay ${relayName} pulsed via dashboard`,
        });
        broadcastCommand(command);

        res.json({
          ok: true,
          correlationId,
          relay: {
            id: relayId,
            name: relayName,
            type: "momentary",
            state: false,
            deviceId: targetDeviceId,
            location: targetLocation,
          },
        });
      } else {
        // Switch: existing behavior
        const { state } = req.body;
        if (typeof state !== "boolean") {
          res.status(400).json({ ok: false, error: "state must be a boolean" });
          return;
        }

        const correlationId = publishCommand({
          deviceId: targetDeviceId,
          location: targetLocation,
          target: relayId,
          action: "set",
          value: state,
          source: "dashboard",
          reason: `Relay ${relayName} set to ${state ? "ON" : "OFF"} via dashboard`,
        });

        if (!correlationId) {
          res.status(503).json({ ok: false, error: "MQTT client not connected" });
          return;
        }

        const command = insertCommand({
          id: correlationId,
          ts: Date.now(),
          deviceId: targetDeviceId,
          target: relayId,
          action: "set",
          value: state,
          source: "dashboard",
          reason: `Relay ${relayName} set to ${state ? "ON" : "OFF"} via dashboard`,
        });
        broadcastCommand(command);

        updateActuatorState(targetDeviceId, relayId, state);

        res.json({
          ok: true,
          correlationId,
          relay: {
            id: relayId,
            name: relayName,
            state,
            deviceId: targetDeviceId,
            location: targetLocation,
          },
        });
      }
    } catch (err) {
      console.error("Error setting relay state", err);
      res.status(500).json({ ok: false, error: "Failed to set relay state" });
    }
  });

  // PATCH /api/devices/:deviceId/relays/:id - Update relay name
  router.patch("/:id", (req: Request, res: Response) => {
    try {
      const { name } = req.body;
      const relayId = req.params.id as string;
      const targetDeviceId = String(req.params.deviceId);

      // Find the actuator to get its device
      const actuators = getDeviceActuators(targetDeviceId);
      const actuator = actuators.find((a) => a.id === relayId);

      if (!actuator) {
        res.status(404).json({ ok: false, error: "Relay not found" });
        return;
      }

      if (name !== undefined) {
        const updated = updateActuatorName(targetDeviceId, relayId, name);
        if (!updated) {
          res.status(500).json({ ok: false, error: "Failed to update relay name" });
          return;
        }
      }

      // Return updated relay info
      const device = getDevice(targetDeviceId);
      const updatedActuator = device?.capabilities.actuators.find((a) => a.id === relayId);

      res.json({
        ok: true,
        relay: {
          id: relayId,
          name: device?.actuatorNames[relayId] ?? updatedActuator?.name ?? relayId,
          state: updatedActuator?.state ?? false,
          deviceId: targetDeviceId,
          location: device?.location ?? actuator.location,
          deviceOnline: device?.online ?? false,
        },
      });
    } catch (err) {
      console.error("Error updating relay", err);
      res.status(500).json({ ok: false, error: "Failed to update relay" });
    }
  });

  // DELETE /api/devices/:deviceId/relays/:id - Remove custom name (relay itself comes from device)
  router.delete("/:id", (req: Request, res: Response) => {
    try {
      const relayId = req.params.id as string;

      // Find the actuator to get its device
      const actuators = getDeviceActuators(String(req.params.deviceId));
      const actuator = actuators.find((a) => a.id === relayId);

      if (!actuator) {
        res.status(404).json({ ok: false, error: "Relay not found" });
        return;
      }

      // Remove custom name (actuator still exists from device capabilities)
      removeActuatorName(actuator.deviceId, relayId);

      res.json({ ok: true });
    } catch (err) {
      console.error("Error deleting relay", err);
      res.status(500).json({ ok: false, error: "Failed to delete relay" });
    }
  });

  return router;
}

File: apps/cortex/src/api/relays.py
"""
Relay routes: GET/POST/PATCH/DELETE /api/devices/:deviceId/relays/:id

Ported from apps/api/src/routes/relays.ts
"""

import time
import uuid
from typing import Any

from fastapi import APIRouter, Request
from pydantic import BaseModel


class RelayStateRequest(BaseModel):
    state: bool | None = None


class RelayNameRequest(BaseModel):
    name: str | None = None


def create_relays_router(sqlite, ws_server) -> APIRouter:
    r = APIRouter()

    @r.get("")
    async def list_relays(deviceId: str):
        try:
            actuators = sqlite.get_device_actuators(deviceId)
            relays = [
                {
                    "id": a.id,
                    "name": a.custom_name or a.name or a.id,
                    "type": a.type or "switch",
                    "state": a.state if a.state is not None else False,
                    "updatedAt": int(time.time() * 1000),
                    "deviceId": a.device_id,
                    "location": a.location,
                    "deviceOnline": a.device_online,
                }
                for a in actuators
            ]
            return {"ok": True, "relays": relays}
        except Exception as e:
            return {"ok": False, "error": f"Failed to fetch relays: {e}"}

    @r.get("/{relay_id}")
    async def get_relay(deviceId: str, relay_id: str):
        try:
            actuators = sqlite.get_device_actuators(deviceId)
            actuator = next((a for a in actuators if a.id == relay_id), None)
            if not actuator:
                return {"ok": False, "error": "Relay not found"}
            return {
                "ok": True,
                "relay": {
                    "id": actuator.id,
                    "name": actuator.custom_name or actuator.name or actuator.id,
                    "type": actuator.type or "switch",
                    "state": actuator.state if actuator.state is not None else False,
                    "deviceId": actuator.device_id,
                    "location": actuator.location,
                    "deviceOnline": actuator.device_online,
                },
            }
        except Exception as e:
            return {"ok": False, "error": f"Failed to fetch relay: {e}"}

    @r.post("/{relay_id}")
    async def control_relay(deviceId: str, relay_id: str, body: RelayStateRequest, request: Request):
        try:
            actuators = sqlite.get_device_actuators(deviceId)
            actuator = next((a for a in actuators if a.id == relay_id), None)
            if not actuator:
                return {"ok": False, "error": "Relay not found"}

            relay_name = actuator.custom_name or actuator.name or relay_id
            location = actuator.location

            mqtt = getattr(request.app.state, "mqtt", None)

            if actuator.type == "momentary":
                # Momentary: send pulse
                correlation_id = f"cmd-{uuid.uuid4().hex[:8]}"
                now = int(time.time() * 1000)

                envelope = {
                    "v": 1, "ts": now, "correlationId": correlation_id,
                    "source": "dashboard", "deviceId": deviceId, "location": location,
                    "type": "command",
                    "payload": {
                        "target": relay_id, "action": "pulse", "value": True,
                        "reason": f"Relay {relay_name} pulsed via dashboard", "ttl": 30000,
                    },
                }

                if not mqtt or not mqtt.is_connected:
                    return {"ok": False, "error": "MQTT client not connected"}

                mqtt.publish_json(f"home/{location}/{deviceId}/command", envelope)
                cmd = sqlite.insert_command(
                    id=correlation_id, ts=now, device_id=deviceId,
                    target=relay_id, action="pulse", value=True,
                    source="dashboard", reason=f"Relay {relay_name} pulsed via dashboard",
                )
                await ws_server.broadcast_command(sqlite.command_to_dict(cmd))

                return {
                    "ok": True, "correlationId": correlation_id,
                    "relay": {"id": relay_id, "name": relay_name, "type": "momentary",
                              "state": False, "deviceId": deviceId, "location": location},
                }
            else:
                # Switch: set state
                if body.state is None:
                    return {"ok": False, "error": "state must be a boolean"}

                correlation_id = f"cmd-{uuid.uuid4().hex[:8]}"
                now = int(time.time() * 1000)
                reason = f"Relay {relay_name} set to {'ON' if body.state else 'OFF'} via dashboard"

                envelope = {
                    "v": 1, "ts": now, "correlationId": correlation_id,
                    "source": "dashboard", "deviceId": deviceId, "location": location,
                    "type": "command",
                    "payload": {
                        "target": relay_id, "action": "set", "value": body.state,
                        "reason": reason, "ttl": 30000,
                    },
                }

                if not mqtt or not mqtt.is_connected:
                    return {"ok": False, "error": "MQTT client not connected"}

                mqtt.publish_json(f"home/{location}/{deviceId}/command", envelope)
                cmd = sqlite.insert_command(
                    id=correlation_id, ts=now, device_id=deviceId,
                    target=relay_id, action="set", value=body.state,
                    source="dashboard", reason=reason,
                )
                await ws_server.broadcast_command(sqlite.command_to_dict(cmd))

                sqlite.update_actuator_state(deviceId, relay_id, body.state)

                return {
                    "ok": True, "correlationId": correlation_id,
                    "relay": {"id": relay_id, "name": relay_name, "state": body.state,
                              "deviceId": deviceId, "location": location},
                }
        except Exception as e:
            return {"ok": False, "error": f"Failed to set relay state: {e}"}

    @r.patch("/{relay_id}")
    async def update_relay_name(deviceId: str, relay_id: str, body: RelayNameRequest):
        try:
            actuators = sqlite.get_device_actuators(deviceId)
            actuator = next((a for a in actuators if a.id == relay_id), None)
            if not actuator:
                return {"ok": False, "error": "Relay not found"}

            if body.name is not None:
                if not sqlite.update_actuator_name(deviceId, relay_id, body.name):
                    return {"ok": False, "error": "Failed to update relay name"}

            device = sqlite.get_device(deviceId)
            updated = next(
                (a for a in (device.capabilities.actuators if device else []) if a.id == relay_id),
                None,
            )

            return {
                "ok": True,
                "relay": {
                    "id": relay_id,
                    "name": (device.actuator_names.get(relay_id) if device else None)
                           or (updated.name if updated else relay_id),
                    "state": updated.state if updated else False,
                    "deviceId": deviceId,
                    "location": device.location if device else actuator.location,
                    "deviceOnline": device.online if device else False,
                },
            }
        except Exception as e:
            return {"ok": False, "error": f"Failed to update relay: {e}"}

    @r.delete("/{relay_id}")
    async def delete_relay_name(deviceId: str, relay_id: str):
        try:
            actuators = sqlite.get_device_actuators(deviceId)
            actuator = next((a for a in actuators if a.id == relay_id), None)
            if not actuator:
                return {"ok": False, "error": "Relay not found"}

            sqlite.remove_actuator_name(actuator.device_id, relay_id)
            return {"ok": True}
        except Exception as e:
            return {"ok": False, "error": f"Failed to delete relay: {e}"}

    return r
